cmd/argocd/commands/plugin.go

```diff
@@ -0,0 +1,142 @@
+package commands
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// DefaultPluginHandler implements the PluginHandler interface
+type DefaultPluginHandler struct {
+	ValidPrefixes []string
+	lookPath      func(file string) (string, error)
+	run           func(cmd *exec.Cmd) error
+}
+
+// NewDefaultPluginHandler instantiates the DefaultPluginHandler
+func NewDefaultPluginHandler(validPrefixes []string) *DefaultPluginHandler {
+	return &DefaultPluginHandler{
+		ValidPrefixes: validPrefixes,
+		lookPath:      exec.LookPath,
+		run: func(cmd *exec.Cmd) error {
+			return cmd.Run()
+		},
+	}
+}
+
+// HandleCommandExecutionError processes the error returned from executing the command.
+// It handles both standard Argo CD commands and plugin commands. We don't require to return
+// error but we are doing it to cover various test scenarios.
+func (h *DefaultPluginHandler) HandleCommandExecutionError(err error, isArgocdCLI bool, args []string) error {
+	// the log level needs to be setup manually here since the initConfig()
+	// set by the cobra.OnInitialize() was never executed because cmd.Execute()
+	// gave us a non-nil error.
+	initConfig()
+	log.SetLevel(log.DebugLevel)
+	// If it's an unknown command error, attempt to handle it as a plugin.
+	// Unfortunately, cobra doesn't handle this error, so we need to assume
+	// that error consists of substring "unknown command".
+	// https://github.com/spf13/cobra/pull/2167
+	if isArgocdCLI && strings.Contains(err.Error(), "unknown command") {
+		log.Debug("command does not exist, looking for a plugin...")
+		pluginPath, pluginErr := h.handlePluginCommand(args[1:])
+		// IMP: If a plugin doesn't exist, the returned path will be empty along with nil error
+		// This means the command is neither a normal Argo CD Command nor a plugin.
+		if pluginErr != nil {
+			// If plugin handling fails, report the plugin error and exit
+			log.Errorf("Error: %v\n", pluginErr)
+			return pluginErr
```

Comment:
do not log error before returning it. Let the caller handle the error when you are returning it.

```
return fmt.Errorf("Error: %v\n", pluginErr)
```